### PR TITLE
Spatial panning WIP

### PR DESCRIPTION
**Pull request by jvntf**
Added a new class called Spatializer as an interface for spatial panning and listening

This PR : 
- removes Panner3D() from panner.js
- Spatializer can be constructed with a type argument (accepts 'listener' or defaults to 'panner')
- Using the alternate constructors p5.Panner3D() or p5.Listener3D() (prototypal inheritance) yield equivalent results

### CODE REVIEW

**Comment by b2renger** [comment]
@jvntf That's cool ! (aswell as your awesome work on effetcts)
Let me know when you think it's ready for testing, I wanted to work on 3D audio on this flocking algo : https://b2renger.github.io/Introduction_p5js/06_webgl_05/index.html

Can't wait to try all those new features you're adding

**Comment by jvntf** [comment]
@b2renger test away! i've just included a very basic test in the examples folder (examples/spatial_panning)-- i'm not 100% sure that the positionZ movement is sounding as it should. looking forward to hearing it with the flocking alg!

**Comment by b2renger** [comment]
@jvntf there's a small typo in your example : https://github.com/jvntf/p5.js-sound/commit/87218c245c1498f7157a6bf93a08fdd8e13e1206

Hard to tell if the z-position movement is right with this example though. Maybe my experiments with flocking will help us hear better.

There is an issue with the creation of a listener though, I created a pull request on your fork in the spatialPanning branch, if you want to check out the broken example
it's here : https://github.com/jvntf/p5.js-sound/pull/1

**Comment by jvntf** [comment]
The Listener3D object can be constructed now-- It seems to be updating all the audioContext.listener's audio params correctly, but there is no effect on the sound. @b2renger any ideas?

**Comment by b2renger** [comment]
@jvntf great thanks, I've updated the flock example in the pr. The position seems to work, but I can't seem to be able to call the orientation methods.

**Comment by b2renger** [comment]
@jvntf the spatial listenning example is still broken it seems that now we have a problem with position again

I got this : 
" 
p5.sound.js:8223 Uncaught TypeError: Cannot read property 'positionX' of undefined
    at p5.Listener3D.positionX (p5.sound.js:8223)
    at p5.Listener3D.position (p5.sound.js:8207)
    at draw (sketch.js:62)
    at p5.redraw (p5.js:17205)
    at p5.<anonymous> (p5.js:12931)
    at _runIfPreloadsAreDone (p5.js:12848)
    at p5._decrementPreload (p5.js:12855)
    at p5.sound.js:889
    at p5.sound.js:922
"
are those examples running for you ?

**Comment by jvntf** [comment]
thanks @b2renger ! had forgotten to change a variable name when i separated panner3d and listener3d into different files.

**Comment by b2renger** [comment]
@jvntf it looks nice ! I'm playing a bit but the results with the flocking algorithm are a bit disappointing. I'm trying to make each boid have it's own synthesized sound now.

One thing we might miss is the possibility to set those : 

panner.maxDistance 
panner.rolloffFactor

File: src/listener3d.js [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+  p5.Listener3D.prototype.process = function(src) {
+    src.connect(this.input);
+  }
+  /**
+   * Set the X,Y,Z position of the Panner
+   * @param  {[Number]} xVal
+   * @param  {[Number]} yVal
+   * @param  {[Number]} zVal
+   * @param  {[Number]} time
+   * @return {[Array]}      [Updated x, y, z values as an array]
```
Comment (therewasaguy) [inline src/listener3d.js]: square brackets are for optional parameters, like this

`@param  {Number} [time]`

**Comment by therewasaguy** [comment]
looking good, I'm sorry I haven't dug in as deep as I would like to yet. Talking with @jvntf about long term plan for the listener, we should look into integrating Listener3d into the p5.soundOut (and do a better job of exposing that to users). There's only one listener, and only one soundOut, so they can live together.

A possible solution in the meantime is to merge this but leave out documentation for the listener, which we can move to soundOut in a separate branch

**Comment by b2renger** [comment]
Yes I agree with you jason, the listener should be in the p5.soundOut it totally makes sense.

**Comment by jvntf** [comment]
i agree that listener should go to soundOut! for now, this PR commented out the doc of listener3d, and includes a simple 3d panning example, moving a sound through space.

**Comment by therewasaguy** [comment]
let's fix the merge conflicts, merge this, and then I'm up for doing listener as part of p5.soundOut in a separate PR (and I'll ask for your review)

**Pull request merged**